timemachine/cpp/src/neighborlist.hpp

```diff
@@ -25,63 +29,57 @@ template <typename RealType> class Neighborlist {
 
 public:
     // N - number of atoms
-    Neighborlist(const int NC, const int NR = 0);
+    Neighborlist(const int N);
 
     ~Neighborlist();
 
-    std::vector<std::vector<int>>
-    get_nblist_host(const int NC, const double *h_coords, const double *h_box, const double cutoff);
+    void set_row_idxs(std::vector<unsigned int> idxs);
 
-    std::vector<std::vector<int>> get_nblist_host(
-        const int NC,
-        const int NR,
-        const double *h_column_coords,
-        const double *h_row_coords,
-        const double *h_box,
-        const double cutoff);
+    void reset_row_idxs();
 
-    void build_nblist_device(
-        const int NC, const double *d_coords, const double *d_box, const double cutoff, cudaStream_t stream);
+    void reset_row_idxs_device(const cudaStream_t stream);
+
+    void resize(const int size);
+
+    void resize_device(const int size, const cudaStream_t stream);
+
+    void set_idxs_device(
+        const int NC, const int NR, unsigned int *column_idxs, unsigned int *row_idxs, const cudaStream_t stream);
+
+    std::vector<std::vector<int>>
+    get_nblist_host(const int N, const double *h_coords, const double *h_box, const double cutoff);
 
     void build_nblist_device(
-        const int NC,
-        const int NR,
-        const double *d_col_coords,
-        const double *d_row_coords,
-        const double *d_box,
-        const double cutoff,
-        cudaStream_t stream);
+        const int N, const double *d_coords, const double *d_box, const double cutoff, const cudaStream_t stream);
 
     void compute_block_bounds_host(
-        const int NC, const int D, const double *h_coords, const double *h_box, double *h_bb_ctrs, double *h_bb_exts);
+        const int N, const double *h_coords, const double *h_box, double *h_bb_ctrs, double *h_bb_exts);
 
     unsigned int *get_ixn_atoms() { return d_ixn_atoms_; }
 
     int *get_ixn_tiles() { return d_ixn_tiles_; }
 
     unsigned int *get_ixn_count() { return d_ixn_count_; }
 
+    unsigned int *get_row_idxs() { return d_row_idxs_; }
+
+    int get_num_row_idxs() { return NR_; }
+
     // get max number of row blocks
     int num_row_blocks() const;
 
 private:
-    // Indicates that should compute interactions of all rows and columns
-    // when a row is provided, otherwise only compute upper triangle of the ixn matrix.
-    bool compute_full_matrix() const;
+    // Indicates that should only compute the upper triangle of the interactions matrix, otherwise
+    // will compute the entire matrix.
+    bool compute_upper_triangular() const;
```

Comment:
Do we support computer upper triangles when using `row_idxs=NR` and `col_idxs=NC`?

Reply:
We do not. Though could see a case where we do.